docker/Dockerfile

```diff
@@ -1,101 +1,125 @@
-      build-essential \
-      zlib1g-dev \
-      libncurses5-dev \
-      libgdbm-dev \
-      libnss3-dev \
-      libssl-dev \
-      libreadline-dev \
-      libffi-dev \
-      libsqlite3-dev \
-      wget \
-      libbz2-dev \
- && mkdir -p /tmp/python-src \
- && wget -qO- https://www.python.org/ftp/python/3.11.5/Python-3.11.5.tgz | \
-    tar xz --strip-components=1 -C /tmp/python-src \
- && cd /tmp/python-src \
- && ./configure --prefix=/opt/python3.11 \
- && make install
-
-FROM ubuntu:18.04 as toolchain
+#-------------------------------------------------------------------------------
+# Stage 1: Builder
+# Clones the repository and builds the necessary artifacts.
+#-------------------------------------------------------------------------------
+FROM ghcr.io/pulp-platform/snitch_cluster:main AS builder
+
+# Clone the snitch_cluster repository into /repo
+RUN git clone -b fix-deps https://github.com/pulp-platform/snitch_cluster /repo
+
+WORKDIR /repo
+
+# Patch .gitmodules to replace SSH URLs with HTTPS
+RUN sed -i 's|git@github.com:|https://github.com/|g' .gitmodules
+
+# Initialize and update submodules (now using HTTPS)
+RUN git submodule update --init --recursive
+
+RUN make -C target/snitch_cluster verilator
+
+RUN make -C target/snitch_cluster DEBUG=ON sw
+
+WORKDIR /
+
+RUN cd /repo/target/snitch_cluster/work-vlt/riscv-isa-sim \
+  && ./configure --prefix=/opt/snitch-spike \
+  && make install
+
+WORKDIR /src
+
+RUN curl -L -o riscv32-snitch-llvm-ubuntu2204-15.0.0-snitch-0.2.0.tar.gz https://github.com/pulp-platform/llvm-project/releases/download/15.0.0-snitch-0.2.0/riscv32-snitch-llvm-ubuntu2204-15.0.0-snitch-0.2.0.tar.gz
+
+RUN tar -xzf riscv32-snitch-llvm-ubuntu2204-15.0.0-snitch-0.2.0.tar.gz -C /src
+
+# # Combine all build steps into a single RUN command to leverage caching.
+# # If the repo URL or commit hash doesn't change, this layer remains cached.
+# RUN git clone --recursive https://github.com/pulp-platform/snitch_cluster.git . \
+#    && git reset --hard c757830632b7f7cd7d8ed511e5cc1eb662b43b37 \
+#    && git submodule update --init --recursive \
+#    # Build the verilator model and software in the same layer
+#    && cd /src/target/snitch_cluster \
+#    && make verilator \
+#    && make DEBUG=ON sw
+
+#-------------------------------------------------------------------------------
+# Stage 2: Final Toolchain Image
+# Creates the final, lean image by copying artifacts from the builder
+# and installing only necessary runtime dependencies.
+#-------------------------------------------------------------------------------
+FROM ubuntu:22.04 AS toolchain
+
+# Set this environment variable early.
 ENV DEBIAN_FRONTEND=noninteractive
-# Runtime dependencies
-COPY --from=builder /src/target/snitch_cluster/bin/snitch_cluster.vlt /opt/snitch-rtl/bin/snitch_cluster.vlt
+
+# Combine related COPY commands to reduce image layers.
+# Copying parent directories is more efficient than many individual sub-directories.
```

Comment:
I've kept the old `COPY` commands below, to trace what has changed. These will need to be removed once things are working.